File: backend/src/service/ranking_service.py
from typing import Optional, List
from src.schemas.ranking_schema import RankingEntry, RankingResponse
from src.infra.database.repositories import ranking_repo


class RankingService:
    """Serviço para cálculo de rankings.

    Implementação mínima — os testes descrevem o comportamento esperado.
    """

    @staticmethod
    async def get_top_ranking(level: str = "global", code: Optional[str] = None, limit: int = 10) -> RankingResponse:
        """Retorna o top do ranking.

        level: 'global'|'estado'|'cidade'
        code: quando nível for 'estado' ou 'cidade' informa o código (ex.: 'pi' ou 'teresina')
        """
        filter_query = None
        if level == "estado":
            filter_query = {"estado_id": code}
        elif level == "cidade":
            filter_query = {"cidade_id": code}

        # Buscar usuários ordenados por pontos no repositório
        users = await ranking_repo.get_users_sorted_by_points(filter_query=filter_query, limit=limit)

        # map to RankingEntry
        top = [RankingEntry(user_id=u.get("_id") or u.get("id"), name=u.get("name"), cidade_id=u.get("cidade_id"), estado_id=u.get("estado_id"), points=int(u.get("points", 0)), position=i + 1) for i, u in enumerate(users)]

        return RankingResponse(level=level, code=code, total=len(users), top=top)

    @staticmethod
    async def get_user_position(user_id: str, level: str = "global", code: Optional[str] = None) -> int:
        """Retorna a posição (1-based) do usuário no ranking do nível informado. Se não encontrado, retorna -1."""
        filter_query = None
        if level == "estado":
            filter_query = {"estado_id": code}
        elif level == "cidade":
            filter_query = {"cidade_id": code}

        users = await ranking_repo.get_users_sorted_by_points(filter_query=filter_query, limit=None)

        for idx, u in enumerate(users):
            if str(u.get("_id") or u.get("id")) == str(user_id):
                return idx + 1

        return -1

    @staticmethod
    async def refresh_ranking(level: str = "global", code: Optional[str] = None, limit: int = 10) -> RankingResponse:
        """Calcula e salva o ranking no banco, retornando o RankingResponse."""
        doc = await ranking_repo.set_ranking(level=level, code=code, limit=limit)
        # convert top
        top = [RankingEntry(user_id=t.get("user_id"), points=int(t.get("points", 0)), position=i + 1) for i, t in enumerate(doc.get("top", []))]
        return RankingResponse(level=level, code=code, total=len(top), top=top)

    @staticmethod
    async def get_cached_ranking(level: str = "global", code: Optional[str] = None, limit: int = 10) -> RankingResponse:
        """Retorna ranking do armazenamento (cache). Caso não exista, calcula e salva antes de retornar."""
        doc = await ranking_repo.get_ranking(level=level, code=code, limit=limit)
        top = [RankingEntry(user_id=t.get("user_id"), points=int(t.get("points", 0)), position=i + 1) for i, t in enumerate(doc.get("top", []))]
        return RankingResponse(level=level, code=code, total=len(top), top=top)

File: frontend/src/services/rankingService.ts
import api from './api';

/**
 * Interface para entrada de ranking retornada pela API
 */
export interface RankingEntry {
  user_id: string;
  name: string | null;
  cidade_id: string | null;
  estado_id: string | null;
  points: number;
  position: number | null;
}

/**
 * Interface para resposta de ranking da API
 */
export interface RankingResponse {
  level: string; // 'global' | 'estado' | 'cidade'
  code: string | null; // código do estado ou cidade
  cidade_nome: string | null; // nome da cidade (quando level=cidade)
  estado_nome: string | null; // nome do estado (quando level=estado)
  total: number;
  top: RankingEntry[];
}

/**
 * Tipo de ranking disponível
 */
export type RankingLevel = 'global' | 'estado' | 'cidade';

/**
 * Serviço para operações relacionadas ao ranking
 */
class RankingService {
  /**
   * Busca o ranking por nível
   * @param level - Nível do ranking ('global', 'estado', 'cidade')
   * @param code - Código do estado ou cidade (obrigatório para 'estado' e 'cidade')
   * @param limit - Número de resultados no top (padrão: 10)
   */
  async getRanking(
    level: RankingLevel,
    code?: string,
    limit: number = 10
  ): Promise<RankingResponse> {
    const params: Record<string, string | number> = { level, limit };
    if (code) {
      params.code = code;
    }

    const response = await api.get<RankingResponse>('/rankings/', { params });
    return response.data;
  }

  /**
   * Busca a posição do usuário no ranking
   * @param userId - ID do usuário
   * @param level - Nível do ranking ('global', 'estado', 'cidade')
   * @param code - Código do estado ou cidade (obrigatório para 'estado' e 'cidade')
   */
  async getUserPosition(
    userId: string,
    level: RankingLevel,
    code?: string
  ): Promise<number> {
    const params: Record<string, string> = { level };
    if (code) {
      params.code = code;
    }

    const response = await api.get<{ position: number }>(
      `/rankings/position/${userId}`,
      { params }
    );
    return response.data.position;
  }

  /**
   * Força recalculação do ranking (admin/cron)
   * @param level - Nível do ranking ('global', 'estado', 'cidade')
   * @param code - Código do estado ou cidade (obrigatório para 'estado' e 'cidade')
   * @param limit - Número de resultados no top (padrão: 10)
   */
  async refreshRanking(
    level: RankingLevel,
    code?: string,
    limit: number = 10
  ): Promise<RankingResponse> {
    const params: Record<string, string | number> = { level, limit };
    if (code) {
      params.code = code;
    }

    const response = await api.post<RankingResponse>('/rankings/refresh', null, {
      params,
    });
    return response.data;
  }
}

export default new RankingService();
